packages/grafana-ui/src/utils/statsCalculator.ts

```diff
@@ -0,0 +1,326 @@
+// Libraries
+import isNumber from 'lodash/isNumber';
+
+import { TableData, NullValueMode } from '../types/index';
+
+export enum StatID {
+  sum = 'sum',
+  max = 'max',
+  min = 'min',
+  logmin = 'logmin',
+  mean = 'mean',
+  last = 'last',
+  first = 'first',
+  count = 'count',
+  range = 'range',
+  diff = 'diff',
+  delta = 'delta',
+  step = 'step',
+
+  allIsZero = 'allIsZero',
+  allIsNull = 'allIsNull',
+}
+
+export interface ColumnStats {
+  [key: string]: any;
+}
+
+// Internal function
+type StatCalculator = (data: TableData, columnIndex: number, ignoreNulls: boolean, nullAsZero: boolean) => ColumnStats;
+
+export interface StatCalculatorInfo {
+  value: string; // The ID - value maps directly to select component
+  label: string; // The name - label for Select component
+  description: string;
+  alias?: string; // optional secondary key.  'avg' vs 'mean', 'total' vs 'sum'
+
+  // Internal details
+  emptyInputResult?: any; // typically null, but some things like 'count' & 'sum' should be zero
+  standard: boolean; // The most common stats can all be calculated in a single pass
+  calculator?: StatCalculator;
+}
+
+/**
+ * @param ids list of stat names or null to get all of them
+ * @param notFound optional error object that will be filled with the names on unknown stats
+ */
+export function getStatsCalculators(ids?: string[], notFound?: string[]): StatCalculatorInfo[] {
+  if (ids === null || ids === undefined) {
+    return listOfStats;
+  }
+  return ids.reduce((list, id) => {
+    const stat = getById(id);
+    if (stat) {
+      list.push(stat);
+    } else if (notFound && id) {
+      notFound.push(id);
+    }
+    return list;
+  }, new Array<StatCalculatorInfo>());
+}
+
+export interface CalculateStatsOptions {
+  data: TableData;
+  columnIndex: number;
+  stats: string[]; // The stats to calculate
+  nullValueMode?: NullValueMode;
+}
+
+/**
+ * @returns an object with a key for each selected stat
+ */
+export function calculateStats(options: CalculateStatsOptions): ColumnStats {
```

Comment:
I'm working on migrating Legend components to grafana/ui now, your work here seems related to what is displayed in graph/`other potential visualizations` legend. I think stats calculation could/should? be independent from series model and maybe calculated from within React components when necessary. My thinking here is that maybe `calculateStats` could be memoized by default to save some cost of stats calculation?

Reply:
Yes -- but I think the `calculateStats` part should happen outside any reusable legend component.  That should probably get passed an already formatted list of stats.

I toyed with putting the stats on the table column, so it could easily be reused -- we should keep this in mind, but no need to do that just yet.

Reply:
I agree. My idea for Legend is to actually be able to set a set of stats it requires, pas it up to the panel, and it's up to the panel controller itself to pass the computed stats back to the legend.

Reply:
👍